cluster_library.c

```diff
@@ -2747,42 +2707,148 @@ int mbulk_resp_loop_assoc(RedisSock *redis_sock, zval *z_result,
-    smart_str hash = {0};
-    zend_string *zstr;
+/* Free an array of zend_string seeds */
+void free_seed_array(zend_string **seeds, uint32_t nseeds) {
+    int i;
+
+    if (seeds == NULL)
+        return;
+
+    for (i = 0; i < nseeds; i++) 
+        zend_string_release(seeds[i]);
+
+    efree(seeds);
+}
+
+static zend_string **get_valid_seeds(HashTable *input, uint32_t *nseeds) {
+    HashTable *valid;
+    uint32_t count, idx = 0;
     zval *z_seed;
+    zend_string *zkey, **seeds = NULL;
+
+    /* Short circuit if we don't have any sees */
+    count = zend_hash_num_elements(input);
+    if (count == 0)
+        return NULL;
+
+    ALLOC_HASHTABLE(valid);
+    zend_hash_init(valid, count, NULL, NULL, 0);
 
-    ZEND_HASH_FOREACH_VAL(ht, z_seed) {
-        zstr = zval_get_string(z_seed);
+    ZEND_HASH_FOREACH_VAL(input, z_seed) {
+        ZVAL_DEREF(z_seed);
+
+        if (Z_TYPE_P(z_seed) != IS_STRING) {
+            php_error_docref(NULL, E_WARNING, "Skipping non-string entry in seeds array");
+            continue;
+        } else if (strrchr(Z_STRVAL_P(z_seed), ':') == NULL) {
+            php_error_docref(NULL, E_WARNING,
+                "Seed '%s' not in host:port format, ignoring", Z_STRVAL_P(z_seed));
+            continue;
+        }
+
+        /* Add as a key to avoid duplicates */
+        zend_hash_str_update_ptr(valid, Z_STRVAL_P(z_seed), Z_STRLEN_P(z_seed), NULL);
+    } ZEND_HASH_FOREACH_END();
+
+    /* We need at least one valid seed */
+    count = zend_hash_num_elements(valid);
+    if (count == 0)
+        goto cleanup;
+
+    /* Populate our return array */
+    seeds = emalloc(sizeof(*seeds) * count);
+    ZEND_HASH_FOREACH_STR_KEY(valid, zkey) {
+        seeds[idx++] = zend_string_dup(zkey, 0);
+    } ZEND_HASH_FOREACH_END();
+
+    *nseeds = idx;
+
+cleanup:
+    zend_hash_destroy(valid);
+    FREE_HASHTABLE(valid);
+
+    return seeds;
+}
+
+/* Validate cluster construction arguments and return a sanitized and validated
+ * array of seeds */
+zend_string**
+cluster_validate_args(double timeout, double read_timeout, HashTable *seeds, 
+                      uint32_t *nseeds, char **errstr) 
+{
+    zend_string **retval;
+    
+    if (timeout < 0L || timeout > INT_MAX) {
+        if (errstr) *errstr = "Invalid timeout";
+        return NULL;
+    }
+
+    if (read_timeout < 0L || read_timeout > INT_MAX) {
+        if (errstr) *errstr = "Invalid read timeout";
+        CLUSTER_THROW_EXCEPTION("Invalid read timeout", 0);
+        return NULL;
+    }
+
+    retval = get_valid_seeds(seeds, nseeds);
+    if (retval == NULL) {
+        if (errstr) *errstr = "No valid seeds detected";
+        return NULL;
+    }
+
+    return retval;
+}
+
+/* Helper function to compare to host:port seeds */
+static int cluster_cmp_seeds(const void *a, const void *b) {
+    zend_string *za = *(zend_string **)a;
+    zend_string *zb = *(zend_string **)b;
+    return strcmp(ZSTR_VAL(za), ZSTR_VAL(zb));
```

Comment:
`zend_string_equals`

Reply:
I can't use `zend_string_equals` here because I'm sorting the strings.